sos/collector/__init__.py

```diff
@@ -134,7 +134,13 @@ class SoSCollector(SoSComponent):
         'upload_pass': None,
         'upload_method': 'auto',
         'upload_no_ssl_verify': False,
-        'upload_protocol': 'auto'
+        'upload_protocol': 'auto',
+        'upload_s3_endpoint': None,
+        'upload_s3_region': None,
+        'upload_s3_bucket': None,
+        'upload_s3_access_key': None,
+        'upload_s3_secret_key': None,
+        'upload_s3_object_prefix': None
```

Comment:
Out of curiosity, are *all* these options *required* for an S3 upload? I've only messed with AWS a little bit myself over the years, and I seem to remember some tools using a json file with...some? all? of these items in it and then just having the tool reference the file. Is that a standard design (and should we support that here?) or am I imagining things?

Reply:
> Out of curiosity, are all these options required for an S3 upload? 

Kind of.

## Required
* **upload_s3_bucket** required in any s3 upload.
* **upload_s3_access_key** required in any s3 upload.
* **upload_s3_secret_key** required in any s3 upload.

## Optional
* **upload_s3_endpoint** default is **s3.amazonaws.com**.
  * Could be removed if only supporting Amazon-hosted buckets, blocking all third-party s3 providers.
* **upload_s3_region** defaults to **us-east-1**
  * This default is usually acceptable for all providers, as even third-party providers tend to implement the API with **us-east-1** as the default region.
  * However, best practice is to provide the correct region for a bucket. This avoids first going to **us-east-1** to determine the correct region. Also allows avoiding upload failures during a regional or service outage in **us-east-1** where it cannot be used to determine the correct region.

## Specific to SOS
* **upload_s3_object_prefix** is a convention created by me for sos so that the filename being uploaded can be combined with an object prefix to define the key (including prefixes) to upload (URL path in standard terminology).
  * As there is no direct access to the awscli to specify the key including its prefixes (i.e. full path to the object inside the bucket) **and** because s3 does not honor normal protocol URL conventions. 
  * A standard URL`proto://<host.domain.tld>/path/of/url` vs. an S3 url `proto://<bucket>/object/prefix/key`
    * e.g. `https://www.example.com/sosreports/2024/01/sosreport-myhost-2024-01-08-askljf.tar.xz` is instead split into pieces where the host is `--endpoint-url www.example.com` and the rest of the URL is `s3://sosreports/2024/01/sosreport-myhost-2024-01-08-askljf.tar.xz`.

    
Since the API has specific differences for how the URL is defined I did not want to hack something together from `--upload-url` which would be unique to sos and not used in awscli or any other s3 client implementations I've seen.

---

> I've only messed with AWS a little bit myself over the years, and I seem to remember some tools using a json file with...some? all? of these items in it and then just having the tool reference the file. Is that a standard design (and should we support that here?) or am I imagining things?

When using **awscli** I've used the `~/.aws/credentials` and the `~/.aws/config` for SSO. Both configuration files are in INI syntax styles. I have also seen custom scripts read from a JSON file, but I do not believe that is a standard design from a review of the [boto3 credentials docs](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html). 

That said I did not test for values coming from either configuration file at all. I presumed they would likely not be setup for sreport submissions and therefore put options for each potential argument required.

----

Reply:
Leaving this conversation as unresolved just in case my response drives additional suggestions.

Reply:
Works for me. If we find that we need or want to add support for getting these values from a json file we can add that later - and in the meantime we can just direct users to putting these values in `sos.conf` if truly desired.

Reply:
Sounds good. If we add a JSON file later I'll also add the `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY` and `AWS_DEFAULT_REGION` environment variables as well as an `--upload-s3-profile` similar to `aws --profile` to override the environment vars by leveraging the profiles in `~/.aws/config` or `~/.aws/credentials`.